music_assistant/server/providers/snapcast/__init__.py

```diff
@@ -0,0 +1,305 @@
+        player = self.mass.players.get(player_id, raise_unavailable=False)
+        if not player:
+            player = Player(
+                player_id=player_id,
+                provider=self.domain,
+                type=PlayerType.PLAYER,
+                name=snap_client.friendly_name,
+                available=True,
+                powered=snap_client.connected,
+                device_info=DeviceInfo(),
+                supported_features=(
+                    PlayerFeature.SYNC,
+                    PlayerFeature.VOLUME_SET,
+                    PlayerFeature.VOLUME_MUTE,
+                ),
+            )
+        self.mass.players.register_or_update(player)
+        player.name = snap_client.friendly_name
+        player.volume_level = snap_client.volume
+        player.volume_muted = snap_client.muted
+        player.available = snap_client.connected
+        player.can_sync_with = tuple(
+            x.identifier for x in self._snapserver.clients if x.identifier != player_id
+        )
+        player.synced_to = self._synced_to(player_id)
+        self.mass.players.register_or_update(player)
+
+    async def unload(self) -> None:
+        """Handle close/cleanup of the provider."""
+        for client in self._snapserver.clients:
+            await self.cmd_stop(client.identifier)
+        self._snapserver.stop()
+
+    async def cmd_volume_set(self, player_id: str, volume_level: int) -> None:
+        """Send VOLUME_SET command to given player."""
+        self.mass.create_task(
+            self._snapserver.client_volume(player_id, {"percent": volume_level, "muted": False})
+        )
+
+    async def cmd_play_url(
+        self,
+        player_id: str,
+        url: str,
+        queue_item: QueueItem | None,  # noqa: ARG002
+    ) -> None:
+        """Send PLAY URL command to given player.
+
+        This is called when the Queue wants the player to start playing a specific url.
+        If an item from the Queue is being played, the QueueItem will be provided with
+        all metadata present.
+
+            - player_id: player_id of the player to handle the command.
+            - url: the url that the player should start playing.
+            - queue_item: the QueueItem that is related to the URL (None when playing direct url).
+        """
+        await self.cmd_stop(player_id)
+
+        stream = self._get_snapstream(player_id)
+        player = self.mass.players.get(player_id, raise_unavailable=False)
+
+        if stream.path != "":
+            group = self._get_snapgroup(player_id)
+            stream_id = await self._get_empty_stream(player_id)
+            await group.set_stream(stream_id)
+
+        stream = self._get_snapstream(player_id)
+
+        stream_host = stream._stream.get("uri").get("host")
+        ffmpeg = (
+            FFmpeg()
+            .option("y")
+            .input(url)
+            .output(f"tcp://{stream_host}", f="u16le", acodec="pcm_s16le", ac=2, ar=48000)
```

Comment:
does this need to be configurable for other sample rates ?

Reply:
This is the default sample rates in snapcast when creating a stream. I will check if I have a way to get it from the api.

Reply:
ah ok all clear then. For some future iteration we might want to skip the intermediate ffmpeg in between and just forward the raw pcm bytes.